pkg/aggregator/ckey/key.go

```diff
@@ -27,49 +23,93 @@ import (
+// nor did benchmarks with xxhash (slightly slower).
 type ContextKey uint64
 
-// KeyGenerator generates key
-// Not safe for concurrent usage
-type KeyGenerator struct {
-	buf []byte
-}
-
 // NewKeyGenerator creates a new key generator
 func NewKeyGenerator() *KeyGenerator {
 	return &KeyGenerator{
-		buf: make([]byte, 0, 1024),
+		seen:  make([]uint64, 512, 512),
+		empty: make([]uint64, 512, 512),
 	}
 }
 
+// KeyGenerator generates hash for the given name, hostname and tags.
+// The tags don't have to be sorted and duplicated tags will be ignored while
+// generating the hash.
+// Not safe for concurrent usage.
+type KeyGenerator struct {
+	// reused buffer to not create a uint64 on the stack every key generation
+	intb uint64
+
+	// seen is used as a hashset to deduplicate the tags when there is more than
+	// 16 and less than 512 tags.
+	seen []uint64
+	// empty is an empty hashset with all values set to 0, to reset `seen`
+	empty []uint64
+
+	// idx is used to deduplicate tags when there is less than 16 tags (faster than the
+	// hashset) or more than 512 tags (hashset has been allocated with 512 values max)
+	idx int
+}
+
 // Generate returns the ContextKey hash for the given parameters.
 // The tags array is sorted in place to avoid heap allocations.
 func (g *KeyGenerator) Generate(name, hostname string, tags []string) ContextKey {
-	g.buf = g.buf[:0]
+	g.intb = 0xc6a4a7935bd1e995
+	g.intb = g.intb ^ murmur3.StringSum64(name)
+	g.intb = g.intb ^ murmur3.StringSum64(hostname)
 
-	// Sort the tags in place. For typical tag slices, we use
-	// the in-place insertion sort to avoid heap allocations.
-	// We default to stdlib's sort package for longer slices.
-	// See `pkg/util/sort.go` for info on the threshold.
-	if len(tags) < util.InsertionSortThreshold {
-		util.InsertionSort(tags)
+	// there is two implementations used here to deduplicate the tags depending on how
+	// many tags we have to process:
+	//   -  16 < n < 512: 		we use a hashset of 512 values. This size has been
+	// 							selected to have space for approximately 500 tags
+	// 							since it nos impacting much the performances,
+	// 							even if the backend is truncating after 100.
+	//   -  n < 16 or n > 512: 	we use a simple for loops, which is faster than
+	//                         	the hashset when there is less than 16 tags, and
+	//                         	we use it as fallback when there is more than 512
+	//                         	because it is the maximum size the allocated
+	//                         	hashset can handle.
+	if len(tags) > 16 && len(tags) < 512 {
+		copy(g.seen, g.empty) // reset the `seen` hashset
+	OUTER:
+		for i := range tags {
+			h := murmur3.StringSum64(tags[i])
+			j := h & 511 // address this hash into the hashset
+			for {
+				if g.seen[j] == 0 { // not seen, we will add it to the hash
+					g.seen[j] = h
+					break
+				} else if g.seen[j] == h {
```

Comment:
I think this is debatable since 1. we rely on murmur3 for the rest of the hash anyway, 2. comparing strings must be slower than `uint64`, even if this branch of the code should happen extremely rarely as you mentioned (only in case of collisions)...
I'll check that we never enter that branch using #8526 and switch to string comparison if that's the case.
WDYT?

Reply:
1. I hope eventually we will fix that as well :)
2. String comparison is slower, but it should be infrequent enough to not matter; we can always compare both hashes and strings.

Reply:
Fair enough :) Let me change that.

Reply:
@vickenty since many of the benchmarks and tests have been done with this implementation and I would want this change to go in `7.30.0`, I've added a comment for a follow-up PR to implement that algorithm with string comparisons instead of hash comparisons. Hope this is enough for now.